Support for UserInfo response encrypted

giacomoa opened the pull request:
Close #10517

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

giacomoa committed:
Support for UserInfo response encrypted

Close #10517

tnorimat commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:
```diff
@@ -25,14 +25,17 @@
-import org.keycloak.crypto.SignatureProvider;
-import org.keycloak.crypto.SignatureSignerContext;
-import org.keycloak.crypto.SignatureVerifierContext;
+import org.keycloak.crypto.*;
```
By following the keycloak's convention, it might be better not to include wildcard "*" in an import directive.

giacomoa replied:
Hello @tnorimat, thanks for your feedback.
My IDE (IntelliJ IDEA) seems to include wildcard automatically when imported class for a package is >=  5, I see there are other wildcard in the project with same convention (e.g org.keycloak.models.* in org.keycloak.models.cache.infinispan.RealmAdapter, java.util.* in org.keycloak.storage.ldap.idm.query.internal.LDAPQuery).
Let me know if I need to provide a fix anyway or if you have further considerations.

tnorimat reviewed (changes requested):
@giacomoa Thank you for our PR. I've not yet investigated this PR's codes completely but added review comments. Could you check them?

giacomoa commented:
Hello @tnorimat , do you have any further feedback about this PR?

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/OIDCClientRegistrationTest.java:
```diff
@@ -321,6 +321,47 @@ public void testMtlsHoKTokenEnabled() throws Exception {
+            OIDCClientRepresentation clientRep = createRep();
+
+            response = reg.oidc().create(clientRep);
+            Assert.assertEquals(Boolean.FALSE, response.getTlsClientCertificateBoundAccessTokens());
+            Assert.assertNotNull(response.getClientSecret());
```
It might be better to delete this two lines.

giacomoa replied:
Ok, done

tnorimat reviewed (commented):
@giacomoa I've added some review comments. Could you check them?

giacomoa committed:
Support for UserInfo response encrypted

Close #10517

giacomoa commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:
```diff
@@ -245,12 +253,22 @@ private Response issueUserInfo(String tokenString) {
             event.detail(Details.SIGNATURE_ALGORITHM, cfg.getUserInfoSignedResponseAlg().toString());
         } else {
-            responseBuilder = Response.ok(claims).header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON);
+            if(cfg.isUserInfoEncryptionRequired()) {
```
Ok, I've done both in new commit

giacomoa commented on services/src/main/java/org/keycloak/protocol/oidc/OIDCAdvancedConfigWrapper.java:
```diff
@@ -68,6 +68,26 @@ public boolean isUserInfoSignatureRequired() {
+        setAttribute(OIDCConfigAttributes.USER_INFO_ENCRYPTED_RESPONSE_ENC, algorithm);
+    }
+
+    public boolean isUserInfoEncryptionRequired() {
```
You're right, I've modified as you suggested in OIDCAdvancedConfigWrapper and added an overload in DefaultTokenManager to manage a default value for encryption_enc value.
I think also ID Token encryption has the same issue, cause it doesn't default encryption_enc to A128CBC-HS256 in encryption_alg is specified. We should manage this in a different bug, do you agree?

tnorimat replied:
Yes. I agree with you.

giacomoa commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/UserInfoEndpoint.java:
```diff
@@ -260,6 +278,35 @@ private Response issueUserInfo(String tokenString) {
+
+        KeyWrapper keyWrapper = PublicKeyStorageManager.getClientPublicKeyWrapper(session, client, JWK.Use.ENCRYPTION, algAlgorithm);
+        if (keyWrapper == null) {
+            throw new RuntimeException("can not get encryption KEK");
```
Ok, done

giacomoa commented on themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:
```diff
@@ -404,6 +404,10 @@ id-token-encrypted-response-enc=ID Token Encryption Content Encryption Algorithm
+user-info-encrypted-response-alg=User Info Response Encryption Key Management Algorithm
+user-info-encrypted-response-alg.tooltip=JWA Algorithm used for key management in encrypting User Info Endpoint responses. This option is needed if you want encrypted User Info Endpoint responses. If left empty, User Info Endpoint responses are not encrypted.
+user-info-encrypted-response-enc=User Info Response Encryption Content Encryption Algorithm
+user-info-encrypted-response-enc.tooltip=JWA Algorithm used for content encryption in encrypting User Info Endpoint responses. This option is needed just if you want encrypted User Info Endpoint responses. If left empty, User Info Endpoint responses are not encrypted.
```
Updated the tooltip according to specification and new implementation provided

tnorimat commented:
@giacomoa I've missed the following 2 points in my code review. Could you also check them?

1. Server Metadata
The following [OIDC server metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) should be returned:
- userinfo_encryption_alg_values_supported
- userinfo_encryption_enc_values_supported 
The following PR might be helpful:
- implementation: https://github.com/keycloak/keycloak/pull/6237/files#diff-11c6b1f6947fb113d3e1c41a019acf8ea908c3c7ce9106c42f46af9ce7aee770
- test: https://github.com/keycloak/keycloak/pull/6237/files#diff-50bce047484fc40b152c965bcbb0be2498df27c36a2aed6c28a04eff0a9d24d8

2. Test Patterns
It might be better to test many combinations of `alg` and `enc` as possible.
The following PR might be helpful.
https://github.com/keycloak/keycloak/pull/6237/files#diff-936e7264188091ab29ccb4bfd96d0f3b4b533ea05a6700596f054a75800f73cb

tnorimat commented:
@giacomoa BTY, do you know use cases that requires UserInfo response encryption? If so, could you tell me them?

giacomoa commented:
> @giacomoa I've missed the following 2 points in my code review. Could you also check them?
> 
> 1. Server Metadata
>    The following [OIDC server metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) should be returned:
> 
> * userinfo_encryption_alg_values_supported
> * userinfo_encryption_enc_values_supported
>   The following PR might be helpful:
> * implementation: https://github.com/keycloak/keycloak/pull/6237/files#diff-11c6b1f6947fb113d3e1c41a019acf8ea908c3c7ce9106c42f46af9ce7aee770
> * test: https://github.com/keycloak/keycloak/pull/6237/files#diff-50bce047484fc40b152c965bcbb0be2498df27c36a2aed6c28a04eff0a9d24d8
> 
> 2. Test Patterns
>    It might be better to test many combinations of `alg` and `enc` as possible.
>    The following PR might be helpful.
>    https://github.com/keycloak/keycloak/pull/6237/files#diff-936e7264188091ab29ccb4bfd96d0f3b4b533ea05a6700596f054a75800f73cb
Hi @tnorimat, I've already implemented Server Metadata response in PR provided here: https://github.com/giacomoa/keycloak/blob/feature/user-info-encrypted/services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProvider.java#L145
is there something missing?
For second point I will provide more tests according to PR you linked.

giacomoa commented:
> @giacomoa BTY, do you know use cases that requires UserInfo response encryption? If so, could you tell me them? I think the motivation of this PR is important.

I added a short motivation in linked issue: https://github.com/keycloak/keycloak/issues/10517
In our company we have requirements to encrypt UserInfo responses for OIDC Client external to our organization to reduce possible sensitive information exposures (e.g in client's server logs)

tnorimat commented:
@giacomoa I'm sorry, I've missed that you've already supported Server Metadata support in this PR.

giacomoa committed:
Support for UserInfo response encrypted

Close #10517

giacomoa commented:
Ok, no problem :) I've updated UserInfoTest to check all combination of alg and enc in last commit.
Let me know if there's something more.

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:
```diff
@@ -246,6 +262,219 @@ public void testSuccess_postMethod_header_textEntity() throws Exception {
+
+    @Test
+    public void testSuccessEncryptedResponseAlgRSA_OAEPEncA256GCM() throws Exception {
+        testUserInfoSignatureAndEncryption(null, JWEConstants.RSA_OAEP, JWEConstants.A256GCM);
```
18 of 19 tests set null to "alg" of signing. I think most use cases use signed  and encrypt. It seems good to set some valid signature algorithm to "alg" of signing in 18 of 19 tests and null to "alg" of signing in remaining test.

mposolda replied:
+1 to this point

Also regarding test coverage, it is great that your PR has good test coverage. On the other hand, 19 new test methods in the UserInfoTest looks like quite a lot to me. Every new test method adds some overhead when running the automated tests... Is it possible to slightly limit this for example to 5-7 methods instead of 19? Ideally including the possibilities for "null" values correctly resolved to expected defaults as Takashi suggested. IMO it is not strictly needed to test every possible combination, but for example just some "representative" subset of them. WDYT?

tnorimat commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/UserInfoTest.java:
```diff
@@ -246,6 +262,219 @@ public void testSuccess_postMethod_header_textEntity() throws Exception {
+    @Test
+    public void testSuccessEncryptedResponseAlgRSA_OAEPEncA256GCM() throws Exception {
+        testUserInfoSignatureAndEncryption(null, JWEConstants.RSA_OAEP, JWEConstants.A256GCM);
+    }
```
It seems good to add a test that sets null to "enc" to test a default behavior (expects A128CBC-HS256).

giacomoa replied:
Hi @tnorimat , 
about testing various combination of signing and encryption algorithms we have following:

- signing alg: 12 possible algs - 3 HS algs not of interest for our use case (simmetric key not available to clients for signature's check); we'll consider 9 total algs

- encryption alg: 3 possible algs

- encryption enc: 6 possible alg + 1 for default; we'll consider 7 total algs

In tests provided I check all combination of encryption enc and alg (21 tests) and in 18 of them I set also a signing algorithm (each signing alg is used in 2 different test), remaining 3 tests only apply encryption without signing; could be ok?

tnorimat replied:
@mposolda Yes, I agree with you.
@giacomoa Could you follow @mposolda 's comment?

tnorimat reviewed (commented):
@giacomoa I've added some comments.  Could you check them?

giacomoa committed:
Support for UserInfo response encrypted
Added tests to check enc algorithm default and more signature algorithms

Close #10517

mposolda reviewed (changes requested):
@giacomoa Thanks for the contribution! And thanks @tnorimat for the review!

I have only one small note regarding the count of the tests. But besides that, it looks good to me. I hope we can have this PR quickly resolved, so it can be included in Keycloak 18 release.

giacomoa committed:
Support for UserInfo response encrypted.
Limit UserInfo encryption test to six combinations of algs

Close #10517

mposolda reviewed (approved):
@giacomoa Thanks! I am approving your PR.

@tnorimat WDYT? I can merge this once you approve this PR as well (assuming tests will pass).

tnorimat reviewed (commented):
It might be better to squash all 5 commits to 1 commit. AFAIK, 1 commit per PR is keycloak's convention.

mposolda commented:
@tnorimat True. I did that during the merge.
@giacomoa Thanks again!
